### ISSUE TITLE
vagrant ssh only possible after restart network

### ISSUE THREAD

**Issue by sebastian-alfers** [hashicorp/vagrant#391]
Hey,

i can not log into my VM after "vagrant up"

i have to start it in gui-mode, then retart my network adapter "sudo /etc/init.d/networking restart"
after this, my VM gets an ip (v4) address and my mac is able to ssh the VM and do the provisioning.

any idea on this?

same isse as here: http://groups.google.com/group/vagrant-up/browse_frm/thread/e951417f59e74b9c

the box is about 5 days old!

Thank you!
Seb

**Comment by matiasepalacios** [comment]
I came up with a fix that works fine so far for me. 

Disclaimer: this is more a hack than a fix, as it does not solves the real problem, which I have no clue how to fix, however, it works...

Of, so first, you have to make sure your /etc/networks/interface file is 'clean'. 
Mine looks like this:


    # This file describes the network interfaces available on your system
    # and how to activate them. For more information, see interfaces(5).
    
    # The loopback network interface 
    auto lo
    iface lo inet loopback
    
    # The primary network interface
    allow-hotplug eth0
    iface eth0 inet dhcp 
    pre-up sleep 2


Then, once you make sure your file is ready, you just create a simple bash script, put it on /etc/init.d and add it to the startup of the VM:

    #! /bin/sh
    # /etc/init.d/interfaces.sh
    #
    
    ### BEGIN INIT INFO
    # Provides: crap
    # Required-Start: 
    # Required-Stop:
    # Should-Start:
    # Should-Stop:
    # Default-Start: 2 3 4 5
    # Default-Stop: 0 1 6
    # Short-Description: Does Vagrant related crap
    # Description: More crap
    ### END INIT INFO
    
    
    sudo cat /etc/network/interfaces > /etc/network/interfaces.bak
    
And to add it, you just run this command:

    sudo /usr/sbin/update-rc.d interfaces.sh defaults

That will make a backup copy of your /etc/network/interfaces file, that you will copy over again when you shutdown the machine. The script that does it is as simple as the first one:

    #! /bin/sh
    # /etc/init.d/interfaces-at-shutdown.sh
    #
    
    ### BEGIN INIT INFO
    # Provides: crap
    # Required-Start:
    # Required-Stop:
    # Should-Start:
    # Should-Stop:
    # Default-Start: 2 3 4 5
    # Default-Stop: 0 1 6
    # Short-Description: Does Vagrant related crap
    # Description: More crap
    ### END INIT INFO
    
    
    sudo cat /etc/network/interfaces.bak > /etc/network/interfaces
    
and then you create a symbolic link to it on /etc/rc0.d like this: 

    sudo ln -s ../init.d/interfaces-at-shutdown.sh K01interfaces

and that's it. Every time the machine starts, it will backup the file, and every time it shutdown it will reverse the file to it's original state, thus, allowing Vagrant to do it's ssh magic.

**Comment by mikhailov** [comment]
guys, it's all great but looks like hacks, there is a way to avoid such complexity by using VirtualBox built-in DHCP server. https://gist.github.com/mikhailov/740fbfc58767fc495fe2 But of course it doesn't suit you if you need full Vagrant

**Comment by StyxOfDynamite** [comment]
I think shoe-horning all these issues into one has actually had a negative effect. that said here seems to be the only viable place to talk through the issues I'm facing.

##Virtualbox fails to configure virtual network adapters when launched via `vagrant up`

I've literally read everything I can find on the various SSH timeout issues and there seems to be a few different _"causes"_ each with their own _"fixes"_

>None of these identify the actual issue I'm having.

I'm working on a Linux Host, With a Virtualbox Provider and I can't get a single box to launch successfully. Many prepackaged boxes won't launch as my physical [chipset](http://ark.intel.com/products/59836/Intel-Pentium-Processor-B960-2M-Cache-2_20-GHz) doesn't support 64 bit Virtualization (Despite supporting a 64 bit OS)

I've built a Vagrant Box from Scratch. I needed an easily producible, distributable ghost blog installation for theme development. To rule out issues with the created box I've also tried with several 32bit boxes that are already in existence.

###If I launch the box with `vagrant up` at a high level the following things happen.

1 The box is imported from my list of boxes to the current directory
1 Network adapters are initialised
1 The VM boots
1 The VM hangs long enough that vagrant gives up trying to SSH
1 The VM continues to boot after waiting 2 minutes, it waits 60 seconds before waiting an additional 60 seconds and then launches *without* configured virtual network adapters.

At this point I cannot ssh into the box using either 'vagrant ssh' or 'ssh vagrant@127.0.0.1'

Now the box has been imported and the ports configured. I can use the Virtual Box GUI to send the shutdown signal.

###If I then use the VirtualBox GUI to power on the machine the following things happen.

1 The VM boots
1 It uses the saved network configuration from the earlier failed attempt at booting
1 it does not hang
1 I can now use `vagrant ssh | vagrant suspend | vagrant up` without issues.

This to me suggests the issue I'm facing is with how `vagrant up` tells VirtualBox to launch the VM.

The workaround of killing the booted VM and then relaunching it via the GUI is not the end of the world, doesn't involve any hacks to any vagrant files. It does however mean I can't rely on vagrant to provision the machine (not a problem as I purpose built the base image to avoid keep provisioning machines in the same way. I do have to manually add shared folders once the machine has been imported. I can get by with this just fine for the time being but will look more closely at the difference between how the `vagrant up` launches the machine to how VirtualBox launches the machine when I can find time.

**Comment by tigrang** [comment]
If you are using NetworkManger, this worked for me (found it on https://www.virtualbox.org/ticket/13873 last comment by zjwolber)

    To resolve this, I added the following to /etc/NetworkManager/NetworkManager.conf:

        [keyfile]
        unmanaged-devices=interface-name:vboxnet0

The explanation is in the linked ticket.

**Comment by nryoung** [comment]
I am going to put my fix here in case it helps somebody who had the same issue as me.

I received this error when trying to `vagrant ssh` in to my vm after provisioning:

```shell
ssh_exchange_identification: Connection closed by remote host
```
after changing my `synced_folder` setting to:
```ruby
config.vm.synced_folder ".", "/var", type: "nfs"
```
Turns out this folder is owned by `root` thus creating this folder would fail on `vagrant up` and cause the networking portion of the vm to not be configured correctly. When I changed my `synced_folder` to:
```ruby
config.vm.synced_folder ".", "/var/<dir not owned by root>", type: "nfs"
```
where `<dir not owned by root>` is a custom path. This allowed vagrant to provision the VM correctly and thus the networking to come up correctly and `vagrant ssh` worked as expected.

**Comment by Xitsa** [comment]
I have the same problem with virtualbox 5.0.2 and vagrant 1.7.4 on ubuntu 14.04 (all x86).
For some reason eth0/eth1 are bound to ip6.
I've changed in settings of VMs NIC type to Am79C973. After that vagrant connects to VM successfully.

**Comment by alexkart** [comment]
Thanks @Xitsa it works for me too

**Comment by timmackinnon** [comment]
Thought I would add some details around the solution that I employed to work around this issue. This appeared on my end to be bound to the existence of a private network ... so something like this in my Vagrantfile:

`config.vm.network "private_network", ip: "192.168.42.2"`

The end result was mixed up `eth0` and `eth1` interfaces that when running with the VBox GUI enabled and logging in, looked something like the following:

`[root@host0 ~]# nmcli connection show
NAME     UUID                                  TYPE            DEVICE  
docker0  df9293c9-051e-4005-81a1-f08a4e9fdccf  bridge          docker0 
eth0     5fb06bd0-0bb0-7ffb-45f1-d6edd65f3e03  802-3-ethernet  eth0    
[root@host0 ~]# nmcli dev status
DEVICE   TYPE      STATE      CONNECTION 
docker0  bridge    connected  docker0    
eth1     ethernet  connected  eth0       
eth0     ethernet  disconnected  --         
lo       loopback  unmanaged  —`

My solution, and it is kind of hacky was to employ a `systemd` service to clean this up at boot time. It looks like the following:

`[Unit]
Description=Service to clean up eth0 and eth1
Wants=network-online.target
After=network.target network-online.target

[Service]
Type=simple
ExecStartPre=-/usr/bin/echo "Network clean up started"
ExecStartPre=-/usr/bin/sleep 5
ExecStartPre=-/usr/bin/nmcli connection delete eth0
ExecStart=/usr/bin/nmcli connection add type ethernet ifname eth0 con-name eth0
ExecStartPost=-/usr/bin/echo "Network clean up completed"
[Install]
WantedBy=multi-user.target` 

and to push my private network config up to the Vagrant layer:

`config.vm.network "private_network", ip: "192.168.42.2", auto_config: false
config.vm.provision "shell",
  run: "always",
  inline: "ifconfig eth1 192.168.42.2 netmask 255.255.255.0 up"`

When running the following:

`#!/bin/bash
for i in {1..100}
do
   echo "#### Iteration $i started ####"
   vagrant up
   vagrant destroy -f
   echo "#### Iteration $i completed ####"
done`

I was able to get my VM to come up `100/100` times.
